src/state.rs

```diff
@@ -259,6 +268,227 @@ impl DemandProxyState {
+                    .on_demand_dns_cache_hits
+                    .get_or_create(&labels)
+                    .inc();
+                r
+            }
+            None => {
+                metrics
+                    .as_ref()
+                    .on_demand_dns_cache_misses
+                    .get_or_create(&labels)
+                    .inc();
+                let jh = tokio::spawn(Self::resolve_on_demand_dns(
+                    self.clone(),
+                    workload,
+                    dns_resolver_config.to_owned(),
+                    dns_resolver_opts.to_owned(),
+                ));
+                match jh.await {
+                    Ok(_) => {
+                        trace!(
+                            "system dns async resolution: task finished for {:?}",
+                            &workload_uid
+                        );
+                    }
+                    Err(e) => {
+                        warn!(
+                            "system dns async resolution: error for {:?} is: {:?}",
+                            &workload_uid, e
+                        );
+                        return Err(Error::NoResolvedAddresses(workload_uid));
+                    }
+                };
+                // try to get it again
+                let updated_rdns = state.get_ips_for_hostname(&hostname);
+                match updated_rdns {
+                    Some(rdns) => rdns,
+                    None => {
+                        return Err(Error::NoResolvedAddresses(workload_uid));
+                    }
+                }
+            }
+        };
+
+        // TODO: add more sophisticated routing logic, perhaps based on ipv4/ipv6 support underneath us.
+        // if/when we support that, this function may need to move to get access to the necessary metadata.
+        // Randomly pick an IP
+        // TODO: do this more efficiently, and not just randomly
+        let Some(ip) = rdns.ips.iter().choose(&mut rand::thread_rng()) else {
+            return Err(Error::EmptyResolvedAddresses(workload_uid));
+        };
+        Ok(*ip)
+    }
+
+    fn resolve_on_demand_dns(
+        mut state: DemandProxyState,
+        workload: &Workload,
+        dns_resolver_cfg: ResolverConfig,
+        dns_resolver_opts: ResolverOpts,
+    ) -> JoinHandle<()> {
+        let workload_uid = workload.uid.to_owned();
+        let hostname = workload.hostname.to_owned();
+        tokio::spawn(async move {
+            trace!("dns workload async task started for {:?}", &hostname);
+            let resolver_result =
+                TokioAsyncResolver::new(dns_resolver_cfg, dns_resolver_opts, TokioHandle);
+            if resolver_result.is_err() {
+                warn!(
+                    "system dns async resolution: error creating resolver for workload {} is: {:?}",
+                    &workload_uid, resolver_result
+                );
+                return;
+            }
+            let r = resolver_result.unwrap();
```

Comment:
**question:**  Do you know what the cost for creating a new resolver per call is? Assuming the resolver could be reused is it worth considering initializing the resolver when DemandProxyState is initialized and storing the resolver for use when resolve_on_demand_dns is called to avoid that cost during the resolution?

Reply:
i have not benchmarked it. happy to explore optimizations here if others have recommendations but the tokio resolver needs to be in the same runtime as it's currently executing in (since each call to `resolve_on_demand_dns` invokes a `tokio::spawn` we need to be careful we aren't trying to create new runtimes within a runtime, or sharing memory unsafely across runtime environments)

Reply:
Probably worth getting some idea how expensive it is before we spend any time optimizing. No need to delay the PR over this.